CP-44372: Integrate NRPE UI with backend interface

BengangY opened the pull request:
Added below use case on XenCenter

Enable/Disable NRPE for host
Configure allow_hosts /Enable Debug/Enable SSL_logging
Set threshold for each check
View existing settings for each host
Validation for the input
Same function for the pool

kc284 commented on XenAdmin/Dialogs/PropertiesDialog.cs:
```diff
@@ -317,6 +318,11 @@ private void Build()
                         dialog.ShowDialog(Program.MainWindow);
                     }
                 }
+                if (isHost || isPool)
```
The page should be available only if the user connects to servers that have the NRPE plugin installed. XC is allowed to connect to older servers which do not have the plugin, so we need to somehow check and hide the page.

kc284 commented on XenAdmin/SettingsPanels/NRPEEditPage.cs:
```diff
@@ -0,0 +1,397 @@
+            Clone = clone;
+            if (IsHost)
+            {
+                InitNRPEGeneralConfiguration();
```
These two methods make server calls on the UI thread, which means that if the network is slow the UI will freeze. Please create a new AsyncAction in the XenModel/Actions/NRPE folder (see my commits in https://github.com/kc284/xenadmin/commits/private/bengangy/CP-44371) and move this code in there. This page should be running the async action, hook the Completed event and update the UI on success/failure accordingly, While the action is running, the UI should have some indication that something is happening on the background, so you may want to add a spinner panel saying something like "Retrieving NRPE configurarion..."

kc284 commented on XenAdmin/SettingsPanels/NRPEEditPage.cs:
```diff
@@ -0,0 +1,397 @@
+        {
+            get
+            {
+                return true;
```
It is not guaranteed that the user will change something. They may click OK without any changes in this dialog or they becasue they have changed something in a different panel. In this case we should not run the action to save changes unnecessarily

kc284 commented on XenAdmin/SettingsPanels/NRPEEditPage.cs:
```diff
@@ -0,0 +1,397 @@
+
+        private void CheckDataGridView_BeginEdit(object sender, DataGridViewCellCancelEventArgs e)
+        {
+            DataGridViewCell currentCell = CheckDataGridView.CurrentRow.Cells[e.ColumnIndex];
```
Here and in the following event handler I'm not convinced that checking the colour is the right way to decide to use thewatermark or not. It would be bette if we had a way to check whether the user has modified the value or not, in which case the cell should have the normal colour to make it obvious.

kc284 commented on XenModel/Actions/Host/NRPEUpdateAction.cs:
```diff
@@ -0,0 +1,191 @@
+        private readonly IXenObject Clone;
+
+        public NRPEUpdateAction(IXenObject host, NRPEHostConfiguration nrpeHostConfiguration, NRPEHostConfiguration nrpeOriginalConfig, bool suppressHistory)
+            : base(host.Connection, Messages.ACTION_CHANGE_POWER_ON, Messages.NRPE_ACTION_CHANGING, suppressHistory)
```
Copy paste error Messages.ACTION_CHANGE_POWER_ON,

kc284 reviewed (changes requested):
As above.

kc284 commented:
A general remark: normally for new controls we would add localized resx placeholders, but at the moment we have stopped shipping localized resources on master, so the palceholders are not needed. This is the first PR with missing placeholders, so please ensure it is merged after #3238 which disables the unit tests.

danilo-delbusso commented on XenAdmin/Properties/AssemblyInfo.cs:
```diff
@@ -35,10 +35,10 @@
-[assembly: AssemblyTitle("[XenCenter]")]
-[assembly: AssemblyDescription("[XenCenter]")]
+[assembly: AssemblyTitle("XenCenter")]
+[assembly: AssemblyDescription("XenCenter")]
 [assembly: AssemblyConfiguration("")]
-[assembly: AssemblyProduct("[XenCenter]")]
+[assembly: AssemblyProduct("XenCenter")]
 
```
Looks like these are leftover from building XenCenter with branding? 👀 We should revert them. Same for the value in `XenAdmin.csproj`

danilo-delbusso commented on XenAdmin/SettingsPanels/NRPEEditPage.cs:
```diff
@@ -0,0 +1,516 @@
+
+        private bool IsNRPEConfigurationChanged()
+        {
+            if (_nrpeCurrentConfig.EnableNRPE != _nrpeOriginalConfig.EnableNRPE ||
+                !_nrpeCurrentConfig.AllowHosts.Equals(_nrpeOriginalConfig.AllowHosts) ||
+                _nrpeCurrentConfig.Debug != _nrpeOriginalConfig.Debug ||
+                _nrpeCurrentConfig.SslLogging != _nrpeOriginalConfig.SslLogging)
+            {
+                return true;
+            }
+            foreach (KeyValuePair<string, NRPEHostConfiguration.Check> kvp in _nrpeCurrentConfig.CheckDict)
+            {
+                NRPEHostConfiguration.Check CurrentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
+                if (CurrentCheck != null && OriginalCheck != null
+                    && (!CurrentCheck.WarningThreshold.Equals(OriginalCheck.WarningThreshold)
+                    || !CurrentCheck.CriticalThreshold.Equals(OriginalCheck.CriticalThreshold)))
+                {
+                    return true;
+                }
+            }
+            return false;
```
I'd put this logic in the `Equals` override for `NRPEHostConfiguration`, so they can be reused. then you can just use a nice and simple `Equals` here.

danilo-delbusso reviewed (changes requested):
Haven't done a complete review but noticed a couple of things while skim-reading

kc284 commented on XenAdmin/Dialogs/PropertiesDialog.cs:
```diff
@@ -317,6 +318,12 @@ private void Build()
                         dialog.ShowDialog(Program.MainWindow);
                     }
                 }
+                if (isPoolOrStandalone && Helpers.XapiEqualOrGreater_23_27_0(connection)
```
Is `isPoolOrStandalone` correct here? This means that if someone wants to set different settings for only one host in a multi-host pool, they cannot do it.

BengangY replied:
It's correct. Have confirmed with Enzo that NRPE page will be shown for a pool and standalone host, not shown for a host in a multi-host pool.

kc284 commented:
- If I enter an invalid value in one of the threshold cells (e.g. 120), and click the OK button, I get an error (little red icon in the cell). If I then change the value to something valid (e.g. 70) and click on a different cell, the little red icon should disappear, but it's still there.
- If I enter an invalid value in a cell in the last row, click to a different tab, then click the OK button, it takes me back to the NRPE page, which is correct, however, the row with the invalid value is out of view, one has to scroll down to see what's wrong. Ideally, the row containing the invalid cell should be selected or at least we should scroll into view. If there are more than one row with invalid cells, it should be at least the first invalid row in view.

BengangY commented on XenModel/Actions/NRPE/NRPEUpdateAction.cs:
```diff
@@ -0,0 +1,165 @@
+                    && (!currentCheck.WarningThreshold.Equals(OriginalCheck.WarningThreshold)
+                    || !currentCheck.CriticalThreshold.Equals(OriginalCheck.CriticalThreshold)))
+                {
+                    SetNRPEThreshold(o, currentCheck.Name, currentCheck.WarningThreshold, currentCheck.CriticalThreshold);
```
This Host.call_plugin supports to set only one threshold at a time.
Host.call_plugin takes a parameter thresholdArgDict because checkName, warningThreshold, and criticalThreshold are 3 keys for this call.

kc284 replied:
Ok. Maybe we can consider it as a future enhancement to the plugin.

BengangY commented on XenModel/Actions/NRPE/NRPERetrieveAction.cs:
```diff
@@ -0,0 +1,143 @@
+
+        private void InitNRPEGeneralConfiguration(IXenObject o)
+        {
+            string status = Host.call_plugin(o.Connection.Session, o.opaque_ref, NRPEHostConfiguration.XAPI_NRPE_PLUGIN_NAME,
```
It's not .ini, json, or yaml file. It's a plain text structure, like this:

```
check_cpu warning threshold - 50 critical threshold - 80
check_host_memory warning threshold - 50 critical threshold - 80
check_vgpu_memory warning threshold - 50 critical threshold - 80
...
```
There is only NRPE plugin using this structure now. For it's hard to parse, new plugins in the future (e.g. SNMP) will use json structure instead.

BengangY commented on XenModel/Actions/NRPE/CheckGroup.cs:
```diff
@@ -0,0 +1,267 @@
+        protected DataGridViewTextBoxCell criticalThresholdCell;
+
+        public string Name { get => name; set => name = value; }
+        public string WarningThresholdDefault { get => warningThresholdDefault; set => warningThresholdDefault = value; }
```
After refactoring, they are not used. Have removed them.

BengangY commented on XenModel/Actions/NRPE/CheckGroup.cs:
```diff
@@ -0,0 +1,267 @@
+            InitCheckGroup(name, labelName, DEFAULT_CHECK_WARNING_THRESHOLD, DEFAULT_CHECK_CRITICAL_THRESHOLD);
+        }
+
+        private void InitCheckGroup(string name, string labelName, string warningThresholdDefaultValue, string criticalThresholdDefaultValue)
```
Updated.

BengangY commented on XenModel/Actions/NRPE/CheckGroup.cs:
```diff
@@ -0,0 +1,267 @@
+            InitCheckGroup(name, labelName, warningThresholdDefaultValue, criticalThresholdDefaultValue);
+        }
+
+        public CheckGroup(string name, string labelName)
```
Updated.

BengangY commented on XenModel/XenModel.csproj:
```diff
@@ -57,6 +57,7 @@
     <Reference Include="System.ServiceProcess" />
     <Reference Include="System.Web" />
     <Reference Include="System.Web.Extensions" />
+    <Reference Include="System.Windows.Forms" />
```
Updated.

BengangY commented on XenModel/Actions/NRPE/NRPEUpdateAction.cs:
```diff
@@ -0,0 +1,165 @@
+                        };
+            string result = Host.call_plugin(host.Connection.Session, host.opaque_ref, NRPEHostConfiguration.XAPI_NRPE_PLUGIN_NAME,
+                NRPEHostConfiguration.XAPI_NRPE_SET_THRESHOLD, thresholdArgDict);
+            log.InfoFormat("Execute nrpe {0}, check={1}, w={2}, c={3}, return: {4}",
```
Updated.

BengangY commented on XenModel/Actions/NRPE/NRPEUpdateAction.cs:
```diff
@@ -0,0 +1,165 @@
+            foreach (KeyValuePair<string, NRPEHostConfiguration.Check> kvp in _nrpeHostConfiguration.CheckDict)
+            {
+                NRPEHostConfiguration.Check currentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
```
Updated.

BengangY commented on XenModel/Actions/NRPE/NRPEUpdateAction.cs:
```diff
@@ -0,0 +1,165 @@
+        private void SetNRPEConfigureForPool()
+        {
+            List<Host> hostList = ((Pool) _clone).Connection.Cache.Hosts.ToList();
+            hostList.ForEach(host =>
```
Updated.

BengangY commented on XenModel/Messages.resx:
```diff
@@ -10092,6 +10092,99 @@ When you configure an NFS storage repository, you simply provide the host name o
+    <value>NRPE service is inactive</value>
+  </data>
+  <data name="NRPE_RETRIEVE_FAILED" xml:space="preserve">
+    <value>Retrieve NRPE configuration failed, please check XenCenter logs.</value>
```
Updated.

BengangY commented on XenModel/Messages.resx:
```diff
@@ -10092,6 +10092,99 @@ When you configure an NFS storage repository, you simply provide the host name o
+  <data name="NRPE_ALLOW_HOSTS_PLACE_HOLDER" xml:space="preserve">
+    <value>Comma separated IP address or domain list, e.g. 192.168.1.1, test.domain.com</value>
+  </data>
+  <data name="NRPE_ALLOW_HOSTS_SAME_ADDRESS" xml:space="preserve">
```
Updated.

BengangY commented on XenAdmin/XenAdmin.csproj:
```diff
@@ -9,7 +9,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>XenAdmin</RootNamespace>
-    <AssemblyName>[XenCenter]</AssemblyName>
+    <AssemblyName>XenCenter</AssemblyName>
```
Updated.

BengangY commented:
> * If I enter an invalid value in one of the threshold cells (e.g. 120), and click the OK button, I get an error (little red icon in the cell). If I then change the value to something valid (e.g. 70) and click on a different cell, the little red icon should disappear, but it's still there.
Updated. Now it will check when the datagridview lost focus each time.
> * If I enter an invalid value in a cell in the last row, click to a different tab, then click the OK button, it takes me back to the NRPE page, which is correct, however, the row with the invalid value is out of view, one has to scroll down to see what's wrong. Ideally, the row containing the invalid cell should be selected or at least we should scroll into view. If there are more than one row with invalid cells, it should be at least the first invalid row in view.
Updated. Added a reselection process when error occurs. Code likes:
```
CheckDataGridView.CurrentCell = CheckDataGridView.Rows[0].Cells[0];
CheckDataGridView.CurrentCell = focusCellWhenErrorOccurs;
```
It will select the first row and then select the error row. The reason is that if it only selects the error row, if the user already selected the error row and scrolled down the datagrid and click to a different tab, then click the OK button, it will not focus on the error row. So it needs a switch process.

kc284 commented on XenAdmin/SettingsPanels/NRPEEditPage.cs:
```diff
@@ -0,0 +1,394 @@
+
+        public AsyncAction SaveSettings()
+        {
+            return new NRPEUpdateAction(_clone, _nrpeCurrentConfig, _nrpeOriginalConfig, true);
```
I didn't spot this earlier: why are we suppressing the history? This means that there is not record of the action on the events TabPage. Same for line 195 and NRPERetrieveAction.

kc284 replied:
Hmm, actually we do suppress this for more properties that we save. It probably needs revisiting at a higher level, so it is probably a no-action for this PR.

kc284 commented:
One more question: I noticed that when we disable NRPE and then we renable it, the previous settings are reset and the user has to set up thresholds again. Is this by design? (Asking because sometimes features do remember the settings when disabled).

kc284 commented:
I added a Thread.Sleep() for a few seconds within the Host.call_plugin() method to simulate what happens if, for example, we have a slow network and server requests take while to return. In this case, if I launch the dialog and click the NRPE page, I see an error pop-up:

<img src="https://github.com/xenserver/xenadmin/assets/3705142/84c68d57-4dd9-47f5-a2b1-47a8472160e0" width="400" height="214">

Which means something is not working correctly if we click on the page while the NRPERetrieveAction is still in progress.

BengangY commented:
> One more question: I noticed that when we disable NRPE and then we renable it, the previous settings are reset and the user has to set up thresholds again. Is this by design? (Asking because sometimes features do remember the settings when disabled).

What were your operation steps? There is a scenario, I'm not sure if this is your situation:
1 The current status of NRPE is enabled.
2 Modify some settings, e.g. thresholds, and disable NRPE.
3 Click OK.
Then XC will only send 'disable NRPE' to XS. It doesn't need to send the other settings because the NRPE has already been disabled. This s by design.

BengangY commented:
> I added a Thread.Sleep() for a few seconds within the Host.call_plugin() method to simulate what happens if, for example, we have a slow network and server requests take while to return. In this case, if I launch the dialog and click the NRPE page, I see an error pop-up:
> 
> <img alt="" width="400" height="214" src="https://user-images.githubusercontent.com/3705142/277519539-84c68d57-4dd9-47f5-a2b1-47a8472160e0.png">
> Which means something is not working correctly if we click on the page while the NRPERetrieveAction is still in progress.

This is an issue caused by my removing the default values of the threshold in datagridview. I will fix it today.
